packages/suite/src/actions/suite/metadataActions.ts

```diff
@@ -864,60 +872,84 @@ export const init =
             return false;
         }
 
-        // 2. set device metadata key (master key). Sometimes, if state is not present
-        if (
-            device.metadata.status === 'disabled' ||
-            (device.metadata.status === 'cancelled' && force && device?.connected)
-        ) {
-            dispatch({ type: METADATA.SET_INITIATING, payload: true });
-            await dispatch(setDeviceMetadataKey());
+        if (!force && getState().metadata.error?.[device.state]) {
+            return false;
         }
 
-        // did user confirm labeling on device? or maybe device was not connected
-        // so suite does not have keys and needs to stop here
-        if (getState().device.selectedDevice?.metadata.status !== 'enabled') {
-            // if no, end here
-            dispatch({ type: METADATA.SET_INITIATING, payload: false });
-            dispatch({ type: METADATA.SET_EDITING, payload: undefined });
-
-            return false;
+        dispatch({ type: METADATA.SET_INITIATING, payload: true });
+        if (getState().metadata.error?.[device.state]) {
+            // remove error note about failed migration potentially set in a previous run
+            dispatch({
+                type: METADATA.SET_ERROR_FOR_DEVICE,
+                payload: {
+                    deviceState: device.state,
+                    failed: false,
+                },
+            });
         }
 
-        // if yes, add metadata keys to accounts
-        if (getState().metadata.initiating) {
-            dispatch(syncMetadataKeys());
+        // 1. set metadata enabled globally
+        if (!getState().metadata.enabled) {
+            dispatch(enableMetadata());
         }
 
-        // 3. connect to provider
-        if (
-            getState().device.selectedDevice?.metadata.status === 'enabled' &&
-            !getState().metadata.providers?.length
-        ) {
-            if (!getState().metadata.initiating) {
-                dispatch({ type: METADATA.SET_INITIATING, payload: true });
+        if (!device.metadata?.[METADATA.ENCRYPTION_VERSION]) {
+            const result = await dispatch(
+                setDeviceMetadataKey(device, METADATA.ENCRYPTION_VERSION),
+            );
+            if (!result?.success) {
+                dispatch({ type: METADATA.SET_INITIATING, payload: false });
+                dispatch({ type: METADATA.SET_EDITING, payload: undefined });
+                dispatch({
+                    type: METADATA.SET_ERROR_FOR_DEVICE,
+                    payload: {
+                        deviceState: device.state!,
+                        failed: true,
+                    },
+                });
+                return false;
             }
+        }
+
+        // 3. we have master key. use it to derive account keys
+        dispatch(syncMetadataKeys(device, METADATA.ENCRYPTION_VERSION));
+
+        device = deviceState
```

Comment:
nitpick: maybe this deserves some new selector like `selectDeviceOptionallyByState`? same select using ternary is on multiple places

Reply:
neext time... 😩